### ISSUE TITLE
FreeStyle Libre 3 support

### ISSUE THREAD

**Issue by marekmand** [JohanDegraeve/xdripswift#259]
Hi there!

It would be great to add support for FreeStyle Libre 3.
The new sensors are available to purchase in DE but their official app is very poor one in comparison to XDrip4iOS.
In the official abbot app here is no support for Apple Watch, no Night Scout upload, no such complex alerts as in XDrip4iOS :(
We would like to use the new small sensors FL3 and still be able to see the BG values in apple watch and to have much smarter app :)
Thank you very much!

Marek, a husband of T1D wife :)

**Comment by LiroyvH** [comment]
Is it actually generally available now or still the "get one trial sensor per patient" only?

**Comment by marekmand** [comment]
In Germany FL3 is generally available and can be purchased even without a prescription.

**Comment by gui-dos** [comment]
I was sent a couple of them to test and I bought a third one on eBay.

The BLE characteristics of the Libre 3 behave very differently and I summed up the sequence of the notifications in <https://github.com/gui-dos/DiaBLE/blob/3eb5343353119a3e23d4d9f6157d2ddcb8695789/DiaBLE/Devices/Abbott.swift#L14-L85>.

I tested also whether they support custom NF commands in the range `A0-DF`: <https://github.com/gui-dos/DiaBLE/blob/3eb5343353119a3e23d4d9f6157d2ddcb8695789/DiaBLE/NFC.swift#L441-L453>.

Note that the NFC manufacturer code is not `07` (Texas Instruments) but `7A` (Abbott Diabetes Care) since they are not using the RF430 anymore.

**Comment by idoodler** [comment]
I recently bought a Libre 3 and previously I was using Libre 2.

Is the Libre 2 integration working as the Libre 3 with the Freestyle Libre 3 App where the app continously shows the recent blood sugar, or do we need to scan the the sensor via NFC to obtain the value? I am asking cause its unclear to me if there is any benefit (apart from the size) regarding xDrip integration.

**TLDR;**
Will the Libre 2 sensor continously send blood sugar values to the xDrip App or do I need to constantly scan it with my Phone?

I am also willing to support you guys in adding Libre 3 support to xDrip.

Thanks for this great project 👍

**Comment by gui-dos** [comment]
If you sniff the Bluetooth traffic (I am using *PacketLogger*, which is included in the *Xcode Additional Tools*) you can see that the Libre 2 sends every minute 46 bytes (10 glucose and temperature values) while the Libre 3 only 35 bytes.

However there are two further data characteristics, `0898195A-EF89-11E9-81B4-2A2AE2DBCCE4` and `08981AB8-EF89-11E9-81B4-2A2AE2DBCCE4`, which notify bigger amounts of trend and historic data after writing specific commands on the advertised data read characteristic: clearly the memory storage of the Libre 3 has been greatly enhanced.

I noticed also that the Libre 3 advertised more frequently, one minute instead two minutes, even after a connection is refused.

**Comment by JohanDegraeve** [comment]
@gui-dos the links to the DiaBLE project, does this code fully work for Libre 3 ?

**Comment by gui-dos** [comment]
The Libre 3 doesn’t advertise anymore with “ABBOTT” + its serial number but with a 12-byte string: you can recognize it only from the advertised data service `089810CC-EF89-11E9-81B4-2A2AE2DBCCE4`

However you have first to use a secondary service `0898203A-EF89-11E9-81B4-2A2AE2DBCCE4` to establish a secure connection, similarly to the Gen2 devices like the Libre 2 Sense/US/CA which require to send a “challenge request” and getting back a “session info”: the encoding keys are now clearly not fixed like in the European Libre 2 for XOR-decrypting the data but are dynamically generated at each connection.


I can just replicate the behaviour of the first command you have to send (`0x11`) and the sensor replies correctly. But since it expects authentication data it disconnects after 30 seconds. But after 30 seconds it is already back to accept a new challenge… :-)

**Comment by miguelpruivo** [comment]
Just ordered a Libre 3 and I was wondering the same. I believe that it’s going to be a while until it’s fully available everywhere, nevertheless, it’ll be great to have the support on xdrip5ios.

**Comment by gui-dos** [comment]
I was sent several Libre 3s gathered, I think, from the German xDrip+ community and I bought a couple more on eBay: I summed up my discoveries on the new sensor in https://github.com/gui-dos/DiaBLE/blob/main/DiaBLE/Libre3.swift.

I used the PacketLogger included in the Xcode Additional Tools to sniff the raw Bluetooth packets and I extracted and sorted out many revealing definitions from the decrypted iOS IPA and from the decompiled Android APK.

Most of the native methods to encode the parameter to pass to the "patch control" commands are highly obfuscated by using whiteCryption’s SCP tools and at each connection the ephemeral keys of the secured session are regenerated. The L3 clearly behaves very differently from the Libre 2 EU but also from the “Gen2” Libre Sense US/CA/AU since they introduced totally new BLE characteristics and even a new microcontroller storing larger sets of data.

The data flows are very regular and predictable and you can easily detect and interpret the sequence of the  “security commands” made by one byte and the different notifications of “one-minute” data, recent “historical” data, past “clinical data”, “factory data”, an “event log”, the “patch status” and ECDH “certificate data”… Decrypting them however will be a very hairy if not impossible task, I fear.

Let’s hope Abbot will release their Partner SDK not only for fitness apps but also for medical ones.

**Comment by pvavolas** [comment]
Bad news for users in Germany. Abbott has discontinued Libre 2 sensors and pushing for Libre 3. Has there been any progress here? Any PR in progress that you can share?

**Comment by JohanDegraeve** [comment]
I'm not aware of any progress

**Comment by opsb** [comment]
They've managed to do it on Android now https://github.com/maheini/FreeStyle-Libre-3-patch. It appears to work by patching the freestyle app so that the app transmits updates to a port. They've only made the patched app available though so no idea how it's actually done.

**Comment by balajeek** [comment]
still ?

**Comment by JulienBub** [comment]
https://github.com/j-kaltes/Juggluco it is there for Android

**Comment by opsb** [comment]
I've placed 2 orders for freestyle libre 2s and in both cases the order got changed to libre 3s. It was always a matter of time but it looks like the libre 2s are coming to EOL soon. I'm at the point where I'd be happy to put up $1000 towards someone working on this issue. I started to look for a suitable platform to do this but it seems the best option bountysource no longer exists. A couple of questions:

1. does anyone have any idea who might be able to implement this feature?
2. what's a good platform for putting up a bounty?

**Comment by callms** [comment]
> It was always a matter of time but it looks like the libre 2s are coming to EOL soon. I'm at the point where I'd be happy to put up $1000 towards someone working on this issue

I might be wrong here, It is my understanding that the issue with FSL3 is more legal than technical.

**Comment by opsb** [comment]
Interesting. I know there apps on android that support the libre 3 buy perhaps they're skirting that issue somehow.

**Comment by jstheoriginal** [comment]
> Hi there,
> 
> The issue is already solved by data downloading from libre “cloud” using
> librelinkup.
> 
> All you need is (assuming your sensor is working with the official libre
> app which stores the reading into the libre “cloud”)
> 
> 1) set your xdrip4ios app to the the follower mode and connect librelinkup
> datasource
> 
> 
> 
> 2. to ensure your xdrip app regularly downloads new glucose readings from
> librelinkup, use your libre sensor as heartbeat (Xdrip -> Bluetooth tab).
> The name of the sensor is visible in connecte BT devices.
> 
> 
> 3. Set the keep-alive mode to Heatbeat
> 
> 
> Here you are :)
> 
> Btw, using similar method different sensors can be connected to xdrip4ios
> app. Usually extra nightscout service is needed. Dexcome 7 is also no
> problem to connect to Xdrip. I successfuly connected everything my wife
> wanted to use (libre 1,2,3; dexcome 6,7; syai - currently using).
> 
> If you need more detials a lot is available online - sadly, not in the
> official docs but in the facebook group (
> https://www.facebook.com/groups/853994615056838/).
> 
> I would update and improve the official docs if I had more time (
> https://xdrip4ios.readthedocs.io/en/latest/). Since it is pro bono, my own
> job and family has higher priority. So far, I only share my own build with
> 100 people.
> 
> Regards, Marek
> 
> 
> 
> Dňa ne 10. 8. 2025 o 13:38 Oliver Searle-Barnes ***@***.***>
> napísala(a):
> […](#)

thanks for the instructions @marekmand. i got my Libre 3 Plus working via LibreLinkUp and xDrip. I wasn't able to get the heartbeat to work yet. Are you using a Libre 3 Plus as well? It just gets stuck scanning forever.

**Comment by marekmand** [comment]
Libre 3 only, not plus. What usually helps is to temporarily disabled bluetooth on the original app (libre, syai, …) while pairing with xDrip app.

**Comment by jstheoriginal** [comment]
@paulplant yeah I did see that and tried it multiple times but it just never worked for some reason. 🤷‍♂️ the trick above worked right away though.

**Comment by paulplant** [comment]
OK, no worries but it is impossible that it doesn't work. Manually force-closing the app revokes all bluetooth activity completely until it is re-opened. If you delete the heartbeat and re-add it following the instructions, does it not connect? (it would be great if you could try it quickly)
